src/OrleansSQLUtils/Storage/Provider/AdoNetStorageProvider.cs

```diff
@@ -0,0 +1,594 @@
+            //it means two grains were activated an the other one succeeded in writing its state.
+            //
+            //NOTE: the storage could return also the new and old ETag (Version), but currently it doesn't.
+            if(storageVersion == grainVersion || storageVersion == string.Empty)
+            {
+                //TODO: Note that this error message should be canonical across back-ends.
+                return new InconsistentStateException($"Version conflict ({operation}): ServiceId={serviceId} ProviderName={providerName} GrainType={normalizedGrainType} GrainId={grainId} ETag={grainVersion}.");
+            }
+
+            return null;
+        }
+
+
+        /// <summary>
+        /// Writes a consistent log message from the given parameters.
+        /// </summary>
+        /// <param name="operationProlog">A free form prolog information to log.</param>
+        /// <param name="serviceId">Service Id.</param>
+        /// <param name="providerName">The name of this storage provider.</param>
+        /// <param name="version">The grain version.</param>
+        /// <param name="normalizedGrainType">Grain type without generics information.</param>
+        /// <param name="grainId">The grain ID.</param>
+        /// <param name="exceptionMessage">An optional exception message information to log.</param>
+        /// <returns>A log string to be printed.</returns>
+        private string LogString(string operationProlog, string serviceId, string providerName, string version, string normalizedGrainType, string grainId, string exceptionMessage = null)
+        {
+            const string Exception = " Exception=";
+            return $"{operationProlog}: ServiceId={serviceId} ProviderName={providerName} GrainType={normalizedGrainType} GrainId={grainId} ETag={version}{(exceptionMessage != null ? Exception + exceptionMessage : string.Empty)}.";
+        }
+
+
+        /// <summary>
+        /// Extracts a grain ID as a string and appends the key extension with '#' infix is present.
+        /// </summary>
+        /// <param name="grainReference">The reference from which to extract the ID.</param>
+        /// <returns>The grain ID as a string.</returns>
+        /// <remarks>This likely should exist in Orleans core in more optimized form.</remarks>
+        private static AdoGrainKey GrainIdAndExtensionAsString(GrainReference grainReference)
+        {
+            //Kudos for https://github.com/tsibelman for the algorithm. See more at https://github.com/dotnet/orleans/issues/1905.
+            string keyExtension;
+            AdoGrainKey key;
+            if(grainReference.IsPrimaryKeyBasedOnLong())
+            {
+                key = new AdoGrainKey(grainReference.GetPrimaryKeyLong(out keyExtension), keyExtension);
+            }
+            else
+            {
+                key = new AdoGrainKey(grainReference.GetPrimaryKey(out keyExtension), keyExtension);
+            }
+
+            if(key == null)
```

Comment:
@tsibelman Cheers! You are right. I wonder what's the use of that function getting the key as string then..? :D